analyzer.go

```diff
@@ -19,92 +20,110 @@ type Analyzer struct {
+		return nil, errors.Wrap(err, "retrieving image identifier")
 	}
 
-	data, err := metadata.GetLayersMetadata(image)
+	appMeta, err := metadata.GetLayersMetadata(image)
 	if err != nil {
-		return metadata.AnalyzedMetadata{}, err
+		return nil, errors.Wrap(err, "getting app image metadata")
 	}
 
-	if !a.SkipLayers {
-		for _, buildpack := range a.Buildpacks {
-			bpLayersDir, err := readBuildpackLayersDir(a.LayersDir, buildpack)
-			if err != nil {
-				return metadata.AnalyzedMetadata{}, err
-			}
+	if a.SkipLayers {
+		a.Logger.Infof("Skipping buildpack layer analysis")
+		return &metadata.AnalyzedMetadata{
+			Image:    imageID,
+			Metadata: appMeta,
+		}, nil
+	}
 
-			metadataLayers := data.MetadataForBuildpack(buildpack.ID).Layers
-			for _, cachedLayer := range bpLayersDir.layers {
-				cacheType := cachedLayer.classifyCache(metadataLayers)
-				switch cacheType {
-				case cacheStaleNoMetadata:
-					a.Logger.Infof("Removing stale cached launch layer '%s', not in metadata \n", cachedLayer.Identifier())
-					if err := cachedLayer.remove(); err != nil {
-						return metadata.AnalyzedMetadata{}, err
-					}
-				case cacheStaleWrongSHA:
-					a.Logger.Infof("Removing stale cached launch layer '%s'", cachedLayer.Identifier())
-					if err := cachedLayer.remove(); err != nil {
-						return metadata.AnalyzedMetadata{}, err
-					}
-				case cacheMalformed:
-					a.Logger.Infof("Removing malformed cached layer '%s'", cachedLayer.Identifier())
-					if err := cachedLayer.remove(); err != nil {
-						return metadata.AnalyzedMetadata{}, err
-					}
-				case cacheNotForLaunch:
-					a.Logger.Infof("Using cached layer '%s'", cachedLayer.Identifier())
-				case cacheValid:
-					a.Logger.Infof("Using cached launch layer '%s'", cachedLayer.Identifier())
-					a.Logger.Infof("Rewriting metadata for layer '%s'", cachedLayer.Identifier())
-					if err := cachedLayer.writeMetadata(metadataLayers); err != nil {
-						return metadata.AnalyzedMetadata{}, err
-					}
-				}
+	cacheMeta, err := cache.RetrieveMetadata()
+	if err != nil {
+		return nil, errors.Wrap(err, "retrieving cache metadata")
+	}
+
+	for _, buildpack := range a.Buildpacks {
+		buildpackDir, err := readBuildpackLayersDir(a.LayersDir, buildpack)
+		if err != nil {
+			return nil, errors.Wrap(err, "reading buildpack layer directory")
+		}
+
+		// Restore metadata for launch=true layers.
+		// The restorer step will restore the layer data for cache=true layers if possible or delete the layer.
+		appLayers := appMeta.MetadataForBuildpack(buildpack.ID).Layers
+		for name, layer := range appLayers {
+			identifier := fmt.Sprintf("%s:%s", buildpack.ID, name)
+			if !layer.Launch {
+				a.Logger.Infof("Not restoring metadata for %q, marked as launch=false", identifier)
+				continue
+			}
+			if layer.Build && !layer.Cache {
+				a.Logger.Infof("Not restoring metadata for %q, marked as build=true, cache=false", identifier)
```

Comment:
Maybe this should also be debug level output?

Reply:
Good point, changed to debug.